src/main/java/org/cloudfoundry/promregator/rewrite/CFAllLabelsMetricFamilySamplesEnricher.java

```diff
@@ -24,43 +26,39 @@
 	public static final String LABELNAME_INSTANCE_NUMBER = "cf_instance_number";
 	
 	private static String[] labelNames = new String[] { LABELNAME_ORGNAME, LABELNAME_SPACENAME, LABELNAME_APPNAME, LABELNAME_INSTANCEID, LABELNAME_INSTANCE_NUMBER };
-	
+	private Map<String, String> labelsAndValuesToAdd = new LinkedHashMap<>();
 	public static String[] getEnrichingLabelNames() {
 		return labelNames.clone();
 	}
-	
-	private String orgName;
-	private String spaceName;
-	private String appName;
-	private String instanceId;
 
 	public CFAllLabelsMetricFamilySamplesEnricher(String orgName, String spaceName, String appName, String instanceId) {
-		this.instanceId = instanceId;
-		this.spaceName = spaceName;
-		this.appName = appName;
-		this.orgName = orgName;
+		this.labelsAndValuesToAdd.put(LABELNAME_ORGNAME, orgName);
+		this.labelsAndValuesToAdd.put(LABELNAME_SPACENAME, spaceName);
+		this.labelsAndValuesToAdd.put(LABELNAME_APPNAME, appName);
+		this.labelsAndValuesToAdd.put(LABELNAME_INSTANCEID, instanceId);
+		this.labelsAndValuesToAdd.put(LABELNAME_INSTANCE_NUMBER, getInstanceFromInstanceId(instanceId));
 	}
 	
 	@Override
 	protected List<String> getEnrichedLabelNames(List<String> original) {
 		List<String> clone = new LinkedList<>(original);
-		Collections.addAll(clone, labelNames);
-		
+		Collections.addAll(clone, labelsAndValuesToAdd.keySet().toArray(new String[0]));
 		return clone;
 	}
 	
 	@Override
 	public List<String> getEnrichedLabelValues(List<String> original) {
 		List<String> clone = new LinkedList<>(original);
-		
-		clone.add(this.orgName);
-		clone.add(this.spaceName);
-		clone.add(this.appName);
-		clone.add(this.instanceId);
-		clone.add(getInstanceFromInstanceId(this.instanceId));
-		
+		Collections.addAll(clone, labelsAndValuesToAdd.values().toArray(new String[0]));
 		return clone;
 	}
+
+	@Override
+	protected void removeDuplicateLabels(List<String> original){
+		for(String label: original) {
+			labelsAndValuesToAdd.remove(label);
```

Comment:
Changing the state of the Enricher could be dangerous - for instance due to the coding here https://github.com/promregator/promregator/blob/b55cc3479837fcb1fe5e096d0cb097cd1fb25c4a/src/main/java/org/cloudfoundry/promregator/endpoint/SingleTargetMetricsEndpoint.java#L100

Most likely this won't cause any practical issues by chance, but the approach feels like "asking for trouble" to me.

Perhaps here is a better idea:
* in `getEnrichedLabelNames` check the `original` whether the label's name is already provided there. If yes, skip it (and don't add it to the result, if the name is already there)
* do the same kind of check in `getEnrichedLabelValues` as well. For that, the interface must be extended, such that also the labelNames are handed in as parameters. However, that is safe, as we know all consumers (and can adjust them). 

Then we are also safe with these kind of "dirty cases" I mentioned above.

What do you think?

Reply:
Yea that is a good point.  I will try and implement it this way and add to this pr.

Reply:
perfect. Please feel free to get back to us in case you get stuck somewhere...